src/Features/Core/Completion/MostRecentlyUsedList.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    internal class MostRecentlyUsedList<T>
+    {
+        private readonly object gate = new object();
```

Comment:
It would be quite OK to use _list as the sync object and save an allocation since it never escapes this class.